src/util/live_performance.js

```diff
@@ -0,0 +1,137 @@
+                    obj[max] = Math.max(obj[max] || -Infinity, timer.responseEnd);
+                }
+            }
+        }
+    }
+    return obj;
+}
+
+function getResourceCategory(entry: PerformanceResourceTiming): string {
+    const url = entry.name.split('?')[0];
+
+    // Code may be hosted on various endpoints: CDN, self-hosted,
+    // from unpkg... so this check doesn't include mapbox HTTP URL
+    if (url.includes('mapbox-gl.js')) return 'javascript';
+    if (url.includes('mapbox-gl.css')) return 'css';
+
+    if (isMapboxHTTPFontsURL(url)) return 'font';
+    if (isMapboxHTTPSpriteURL(url)) return 'sprite';
+    if (isMapboxHTTPStyleURL(url)) return 'style';
+    if (isMapboxHTTPTileJSONURL(url)) return 'tilejson';
+
+    return 'other';
+}
+
+function getStyle(resourceTimers: Array<PerformanceEntry>): ?string {
+    if (resourceTimers) {
+        for (const timer of resourceTimers) {
+            const url = timer.name.split('?')[0];
+            if (isMapboxHTTPStyleURL(url)) {
+                const split = url.split('/').slice(-2);
+                if (split.length === 2) {
+                    return `${split[0]}:${split[1]}`;
+                }
+            }
+        }
+    }
+}
+
+export function getLivePerformanceMetrics(data: LivePerformanceData): LivePerformanceMetrics {
+    const resourceTimers = window.performance.getEntriesByType('resource');
+    const resourcesByType = categorize(resourceTimers, getResourceCategory);
+    const counters = getTransferRangePerResourceType(resourcesByType);
+    const devicePixelRatio = window.devicePixelRatio;
+    const connection = window.navigator.connection || window.navigator.mozConnection || window.navigator.webkitConnection;
+    const metrics = {counters: [], metadata: [], attributes: []};
+
+    const addMetric = (arr, name, value) => {
+        if (value) {
+            arr.push({name, value: value.toString()});
+        }
+    };
+
+    for (const counter in counters) {
+        addMetric(metrics.counters, counter, counters[counter]);
+    }
+    if (data.interactionRange[0] !== +Infinity && data.interactionRange[1] !== -Infinity) {
+        addMetric(metrics.counters, "interactionRangeMin", data.interactionRange[0]);
+        addMetric(metrics.counters, "interactionRangeMax", data.interactionRange[1]);
+    }
+
+    addMetric(metrics.attributes, "style", getStyle(resourceTimers));
+    addMetric(metrics.attributes, "terrainEnabled", data.terrainEnabled ? "true" : "false");
+    addMetric(metrics.attributes, "fogEnabled", data.fogEnabled ? "true" : "false");
+    addMetric(metrics.attributes, "projection", data.projection.name);
+
+    addMetric(metrics.metadata, "devicePixelRatio", devicePixelRatio);
+    addMetric(metrics.metadata, "connectionEffectiveType", connection ? connection.effectiveType : undefined);
+    addMetric(metrics.metadata, "cpuCores", window.navigator.hardwareConcurrency);
```

Comment:
I'd suggest keeping the name more similar to the property it's being pulled from. So `navigatorHardwareConcurrency` or `hardwareConcurrency`. It's an abstract value that all browsers don't report the same way. Some browsers report clamped values to maintain privacy. Renaming it obscures that a bit more

I'm also not sure it's going to be very useful so we could drop it. (We used to use this to set the number of workers but ended up switching to a constant number)

Reply:
`hardwareConcurrency` can give a hint on the machine spec, to let us know whether we're on the high-end or low-end spectrum of a machine (for example dual-core mobile would hint a fairly old phone). It's not as good as having true CPU cycle speed, but it's the best we can get.